// agentql-js-common/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseAgentQLError = exports.QuerySyntaxError = void 0;
const AGENTQL_1010_QUERY_SYNTAX_ERROR = 1010;
const SUPPORT_MESSAGE = `
If the issue persists, please reach out for further assistance through support@tinyfish.io or Tiny Fish Discord channel https://discord.gg/agentql.
`.trim();
const REQUEST_ID_MESSAGE = `
Include this Request ID when contacting support for quicker assistance:
`.trim();
class BaseAgentQLError extends Error {
    constructor(message, code = 0, requestId = '', includeFallbackMessage = true) {
        super(message +
            (includeFallbackMessage ? ` ${SUPPORT_MESSAGE}` : '') +
            (requestId ? ` ${REQUEST_ID_MESSAGE} ${requestId}` : ''));
        this.name = this.constructor.name;
        this.code = code;
        this.requestId = requestId;
    }
}
exports.BaseAgentQLError = BaseAgentQLError;
class QuerySyntaxError extends BaseAgentQLError {
    constructor(message, token, row, column) {
        if (!message) {
            message = `Syntax error: unexpected token '${token}' at row ${row}, column ${column}`;
        }
        super(message, AGENTQL_1010_QUERY_SYNTAX_ERROR);
        this.token = token;
        this.row = row;
        this.column = column;
    }
}
exports.QuerySyntaxError = QuerySyntaxError;
