// agentql/dist/core/js-snippets/generated/generated-snippets.js
"use strict";
// Auto-generated file - DO NOT EDIT
Object.defineProperty(exports, "__esModule", { value: true });
exports.snippets = void 0;
exports.snippets = {
    'add-dom-change-listener': `// callback to listen to dom changes
(() => {
  // Check if the observer already exists
  if (!window.domUpdateObserver) {
    const observer = new MutationObserver((mutations) => {
      const now = Date.now();
      window.localStorage.setItem('lastDomChange', now);
    });
    observer.observe(document.body, { childList: true, subtree: true });
    window.domUpdateObserver = observer;
  }
})();
`,
    'generate-accessibility-tree': `/*
 * IMPORTANT NOTICE:
 * This file (generate_accessibility_tree.js) is shared between the repositories 'agentql-client' and 'wadl-inspector'.
 *   agentql-client: /src/agentql/_core/_js_snippets/generate_accessibility_tree.js
 *   wadl-inspector: /app/utils/js_snippets/generate_accessibility_tree.js
 * Any changes made to this file should be mirrored in both repositories to maintain consistency.
 * The only difference is that the version in wadl-inspector contains an \`export\` statement at the end.
 */
function generateAccessibilityTree({
  currentGlobalId,
  processIFrames = true,
  iframePath = '',
  includeHidden = false,
  nodeIdsToIgnore = [],
}) {
  const roleTagMap = {
    role: {
      div: 'generic',
      span: 'text',
      article: 'article',
      header: 'banner',
      footer: 'contentinfo',
      button: 'button',
      th: 'cellheader',
      td: 'cell',
      select: 'listbox',
      dialog: 'dialog',
      ul: 'unorderedlist',
      ol: 'orderedlist',
      figure: 'figure',
      form: 'form',
      table: 'table',
      fieldset: 'group',
      h1: 'heading',
      h2: 'heading',
      h3: 'heading',
      h4: 'heading',
      h5: 'heading',
      h6: 'heading',
      img: 'img',
      nav: 'navigation',
      a: 'link',
      area: 'link',
      li: 'listitem',
      dt: 'term',
      dd: 'listitem',
      main: 'main',
      math: 'math',
      menuitem: 'menuitem',
      meter: 'meter',
      option: 'option',
      progress: 'progressbar',
      section: 'section',
      hr: 'separator',
      textarea: 'textbox',
      tr: 'row',
      tbody: 'rowgroup',
      thead: 'rowgroup',
      tfoot: 'rowgroup',
      aside: 'complementary',
      body: 'document',
      dfn: 'definition',
      dl: 'list',
      mark: 'mark',
      svg: 'img',
    },
    input: {
      button: 'button',
      checkbox: 'checkbox',
      color: 'slider',
      date: 'combobox',
      'datetime-local': 'combobox',
      email: 'textbox',
      file: 'button',
      hidden: 'none',
      image: 'button',
      month: 'combobox',
      number: 'spinbutton',
      password: 'textbox',
      radio: 'radio',
      range: 'slider',
      reset: 'button',
      search: 'searchbox',
      submit: 'button',
      tel: 'textbox',
      text: 'textbox',
      time: 'combobox',
      url: 'textbox',
      week: 'combobox',
    },
  };

  const tfIdMap = new Set();

  function generate_tf_id() {
    ++currentGlobalId;
    tfIdMap.add(currentGlobalId);
    return currentGlobalId;
  }

  function check_and_assign_tfid(node) {
    let nodeId = node.getAttribute('tf623_id');
    if (!nodeId || tfIdMap.has(nodeId)) {
      // if the node does not have a tf_id or the tf_id is already in use, assign a new tf_id
      node.setAttribute('tf623_id', generate_tf_id());
    }
    tfIdMap.add(nodeId);
  }

  function buildAccessibilityTree() {
    const rootNode = document.documentElement;
    const tree = createNode(rootNode, iframePath);
    return tree;
  }

  function extractAttributes(node) {
    const attributes = { html_tag: node.nodeName.toLowerCase() };
    const skippedAttributes = ['style', 'srcdoc'];
    for (let i = 0; i < node.attributes.length; i++) {
      const attribute = node.attributes[i];
      if (!attribute.specified || !skippedAttributes.includes(attribute.name)) {
        attributes[attribute.name] = attribute.value || true;
      }
    }

    // extract the background image attribute separately
    const computedStyle = window.getComputedStyle(node);
    const backgroundImage = computedStyle.getPropertyValue('background-image');
    if (backgroundImage && backgroundImage !== 'none') {
      attributes['background-image'] = backgroundImage;
    }

    return attributes;
  }

  function processHTML(node) {
    for (let child of node.childNodes) {
      const childTag = child.nodeName.toLowerCase();
      if (childTag === 'head') {
        return processHead(child);
      }
    }
    return null;
  }

  function processHead(node) {
    for (let child of node.childNodes) {
      const childTag = child.nodeName.toLowerCase();
      if (childTag === 'title') {
        check_and_assign_tfid(child);
        return {
          role: 'webArea',
          name: child.textContent.trim(),
          attributes: extractAttributes(child),
        };
      }
    }
    return null;
  }

  function getRole(node) {
    const role = node.getAttribute('role') || node.getAttribute('aria-role');
    if (role) {
      return role;
    }
    const tag = node.nodeName.toLowerCase();
    if (tag == 'input') {
      const type = node.getAttribute('type');
      if (type) {
        return roleTagMap.input[type] || type;
      }
      return 'textbox';
    }
    return roleTagMap.role[tag] || 'generic';
  }

  function getName(element) {
    // List of attributes in the order they should be checked
    const attributes = ['aria-label', 'placeholder', 'alt', 'title', 'value', 'name'];

    // Iterate over the attributes and return the first non-empty one
    for (const attribute of attributes) {
      if (element.hasAttribute(attribute)) {
        const value = element.getAttribute(attribute).trim();
        if (value) {
          return value;
        }
      }
    }

    // Fallback to empty string if no non-empty attribute is found
    return '';
  }

  function isElementHidden(element, children) {
    // Check if the element exists
    if (!element) {
      return true;
    }

    //Check if the element is aria-hidden
    if (
      element.getAttribute('aria-hidden') === 'true' ||
      element.getAttribute('aria-hidden') === true
    ) {
      return true;
    }

    const style = window.getComputedStyle(element);

    // Check if the element is hidden by display property
    if (style.display === 'none') {
      return true;
    }

    // Check if the element is hidden by visibility property. If the element has no visible children, it is considered hidden.
    if (style.visibility !== 'visible' && !children) {
      return true;
    }

    // Check if the element is hidden by opacity
    if (style.opacity === '0') {
      return true;
    }

    return false;
  }

  function smartTrim(text) {
    // remove leading and trailing whitespace and replace multiple spaces with a single space
    return text.replace(/^[\s\\n]+|\s+\$/g, ' ');
  }

  function addTextChild(text, parentNode, childNode, childrenArray, iFramePath) {
    const span = document.createElement('span');
    span.textContent = text;
    if (parentNode.contains(childNode)) {
      parentNode.insertBefore(span, childNode);
      parentNode.removeChild(childNode);
    }
    const node = createNode(span, iFramePath);
    if (node) {
      childrenArray.push(node);
    }
  }

  function createNode(node, currentIframePath) {
    if (!node) {
      return null;
    }
    const tag = node.nodeName.toLowerCase();
    const skippedTags = ['script', 'style', 'path', 'g', 'hr', 'br', 'noscript', 'head'];
    if (skippedTags.includes(tag) || nodeIdsToIgnore.includes(node.id)) {
      return null;
    }

    let webAreaNode = null;
    let iframe_path = node.getAttribute('iframe_path');
    if (tag === 'html' && !iframe_path && !currentIframePath) {
      webAreaNode = processHTML(node);
    }

    let role = getRole(node);
    let name = getName(node) || '';
    let newIframePath = currentIframePath;

    let children = [];
    let currentChildNodes = node.childNodes;

    check_and_assign_tfid(node);
    let nodeId = node.getAttribute('tf623_id');

    if (currentIframePath && processIFrames) {
      node.setAttribute('iframe_path', currentIframePath);
    }

    if (node.shadowRoot) {
      const shadowRootChildren = node.shadowRoot.children;
      const childrenNodeList = Array.prototype.slice.call(shadowRootChildren);

      if (childrenNodeList.length > 0) {
        currentChildNodes = Array.from(childrenNodeList);
      } else if (node.shadowRoot.textContent.trim() !== '') {
        name = node.shadowRoot.textContent.trim();
      }
    } else if (tag == 'slot') {
      const assignedNodes = node.assignedNodes();
      if (assignedNodes.length !== 0) {
        currentChildNodes = assignedNodes;
      }
    } else if (tag == 'iframe') {
      const iframeDocument = node.contentDocument;
      if (currentIframePath) {
        newIframePath = currentIframePath + '.' + nodeId;
      } else {
        newIframePath = nodeId;
      }

      if (iframeDocument && processIFrames) {
        if (iframeDocument.body) {
          currentChildNodes = iframeDocument.body.childNodes;
        } else {
          currentChildNodes = iframeDocument.childNodes;
        }
      } else if (node.src) {
        role = 'iframe'; // iframe without contentDocument
      }
    }
    const childNodes = Array.from(currentChildNodes).filter((childNode) => {
      return (
        childNode.nodeType === Node.ELEMENT_NODE ||
        (childNode.nodeType === Node.TEXT_NODE && childNode.textContent.trim() !== '')
      );
    });
    for (let i = 0; i < childNodes.length; i++) {
      let childNode = childNodes[i];
      if (childNode.nodeType === Node.TEXT_NODE) {
        const text = smartTrim(childNode.textContent);
        if (text) {
          if (childNodes.length > 1) {
            addTextChild(text, node, childNode, children, newIframePath);
          } else {
            if (!role || role == 'generic') {
              role = 'text';
            }
            if (!name || (node.tagName && node.tagName.toLowerCase() === 'button')) {
              // If the parent has no name, set name to the child text.
              // Special case: textContent for button elements
              name = text;
            } else {
              // If the parent does have a name attribute, retain the child text as a separate node.
              addTextChild(text, node, childNode, children, newIframePath);
            }
          }
        }
      }
      if (childNode.nodeType === Node.ELEMENT_NODE) {
        const child = createNode(childNode, newIframePath);
        if (child) {
          children.push(child);
        }
      }
    }

    const beforeText = window
      .getComputedStyle(node, '::before')
      .getPropertyValue('content')
      .replaceAll('"', '');
    const afterText = window
      .getComputedStyle(node, '::after')
      .getPropertyValue('content')
      .replaceAll('"', '');

    if (beforeText && beforeText !== 'none') {
      name = beforeText + name;
    }
    if (afterText && afterText !== 'none') {
      name = name + afterText;
    }

    if (children.length === 0) {
      children = undefined;
    }

    if (!role && !name && !children) {
      return null;
    }

    if (!includeHidden && isElementHidden(node, children)) {
      return null;
    }

    let node_dict = {
      role: role || 'generic',
      name,
      attributes: extractAttributes(node),
    };
    if (children) {
      node_dict['children'] = children;
    }
    if (webAreaNode) {
      webAreaNode['children'] = [node_dict];
    }

    return webAreaNode || node_dict;
  }

  return {
    tree: buildAccessibilityTree(),
    lastUsedId: currentGlobalId,
  };
}
`,
    'get-last-dom-change': `// Returns a JavaScript snippet that gets the last DOM change timestamp from the local storage

() => window.localStorage.getItem('lastDomChange');
`,
    'remove-dom-change-listener': `(() => {
  if (window.domUpdateObserver) {
    window.domUpdateObserver.disconnect();
    delete window.domUpdateObserver;
  }
})();
`,
    'set-iframe-path': `// Set the iframe_path attribute for iframe node and its children

function setIframePath({ iframe_path }) {
  function set_iframe_path(node) {
    if (!node) {
      return;
    }

    let currentChildNodes = node.childNodes;

    if (iframe_path) {
      node.setAttribute('iframe_path', iframe_path);
    }

    const childNodes = Array.from(currentChildNodes).filter((childNode) => {
      return (
        childNode.nodeType === Node.ELEMENT_NODE ||
        (childNode.nodeType === Node.TEXT_NODE && childNode.textContent.trim() !== '')
      );
    });
    for (let i = 0; i < childNodes.length; i++) {
      let childNode = childNodes[i];

      if (childNode.nodeType === Node.ELEMENT_NODE) {
        set_iframe_path(childNode);
      }
    }
  }
  set_iframe_path(document.documentElement);
}
`,
};
