// agentql/dist/core/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = void 0;
exports.minifyQuery = minifyQuery;
const winston_1 = require("winston");
const colorOptions = {
    emerg: 'red',
    alert: 'red',
    crit: 'red',
    error: 'red',
    warning: 'yellow',
    notice: 'blue',
    info: 'blue',
    debug: 'green',
};
const logLevels = {
    emerg: 0,
    alert: 1,
    crit: 2,
    error: 3,
    warning: 4,
    notice: 5,
    info: 6,
    debug: 7,
};
const logFormat = winston_1.format.printf(({ level, message, timestamp }) => {
    return `${timestamp} [${level.toUpperCase()}]: ${message}`;
});
const logger = (0, winston_1.createLogger)({
    levels: logLevels,
    level: process.env.DEBUG ? 'debug' : 'notice',
    format: winston_1.format.combine(winston_1.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }), logFormat, winston_1.format.colorize({ all: true, colors: colorOptions })),
    transports: [new winston_1.transports.Console()],
});
exports.logger = logger;
function minifyQuery(query) {
    return query.replace(/\n/g, '\\').replace(/ /g, '');
}
